src/mongo_odm/nvp.hpp

```diff
@@ -21,51 +21,35 @@
-
-// Creates serialize() function
-#define MONGO_ODM_SERIALIZE_KEYS                                                                \
-    template <class Archive>                                                                    \
-    void serialize(Archive& ar) {                                                               \
-        mongo_odm_serialize_recur<Archive, 0,                                                   \
-                                  std::tuple_size<decltype(mongo_odm_mapped_fields())>::value>( \
-            ar);                                                                                \
-    }                                                                                           \
-    template <class Archive, size_t I, size_t N>                                                \
-    typename std::enable_if<(I < N), void>::type mongo_odm_serialize_recur(Archive& ar) {       \
-        auto nvp = std::get<I>(mongo_odm_mapped_fields());                                      \
-        ar(cereal::make_nvp(nvp.name, this->*(nvp.t)));                                         \
-        mongo_odm_serialize_recur<Archive, I + 1, N>(ar);                                       \
-    }                                                                                           \
-                                                                                                \
-    template <class Archive, size_t I, size_t N>                                                \
-    typename std::enable_if<(I == N), void>::type mongo_odm_serialize_recur(Archive&) {         \
-        ;                                                                                       \
-    }
-
-// Register members and create serialize() function
-#define MONGO_ODM_MAKE_KEYS(Base, ...)                \
-    using mongo_odm_wrap_base = Base;                 \
-    constexpr static auto mongo_odm_mapped_fields() { \
-        return std::make_tuple(__VA_ARGS__);          \
-    }                                                 \
-    MONGO_ODM_SERIALIZE_KEYS
-
-// If using the mongo_odm::model, then also register _id as a field.
-#define MONGO_ODM_MAKE_KEYS_MODEL(Base, ...) \
-    MONGO_ODM_MAKE_KEYS(Base, MONGO_ODM_NVP(_id), __VA_ARGS__)
-
-#define MONGO_ODM_KEY(value) mongo_odm::hasCallIfFieldIsPresent<decltype(&value), &value>::call()
+#include <bsoncxx/types.hpp>
+
+#include <mongo_odm/expression_syntax.hpp>
+#include <mongo_odm/macros.hpp>
+#include <mongo_odm/util.hpp>
 
 namespace mongo_odm {
 MONGO_ODM_INLINE_NAMESPACE_BEGIN
 
+template <typename NvpT, typename T>
+class NvpCRTP;
+
+struct current_date_t {
+    constexpr current_date_t() {
+    }
+};
+constexpr current_date_t current_date;
+
 /**
  * An object that represents a name-value pair of a member in an object.
  * It is templated on the class of the member and its type.
  */
 template <typename Base, typename T>
-struct Nvp {
+class Nvp : public NvpCRTP<Nvp<Base, T>, T> {
```

Comment:
I think it'd be good to switch the names here. Can you change this to `class nvp`?